### PR TITLE
fix cpu bnb path

### PR DESCRIPTION
**Pull request by jiqing-feng**
Hi @Rocketknight1 @ArthurZucker @SunMarc @gante 

We have "cpu" in hf_device_map when using bnb model in CPU. The cpu device bnb model should be accepted by transformers because CPU backend has been enabled in BNB. Please take a review, thx!

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Looks good, but do you have an example of code that failed before, that is fixed by this PR?

**Comment by jiqing-feng** [comment]
> Looks good, but do you have an example of code that failed before, that is fixed by this PR?

I run this case in a CPU-only device
```python
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig

model_id = "Felladrin/Llama-68M-Chat-v1"

text = ["I am happy because", "This is"]
tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token
input_ids = tokenizer(text, return_tensors="pt", padding=True)

quantization_config = BitsAndBytesConfig(load_in_8bit=True)

model = AutoModelForCausalLM.from_pretrained(model_id, device_map="auto", quantization_config=quantization_config)
model.generation_config.cache_implementation = "static"
model.generate(**input_ids)
```

Error: `IndexError: list index out of range`

**Comment by jiqing-feng** [comment]
Hi @Rocketknight1 . The script can reproduce this error easily both on AWQ and BNB. Besides, it's not safe to get index without checking the list length.

**Comment by jiqing-feng** [comment]
Hi @Rocketknight1 @ArthurZucker @SunMarc @gante @zucchini-nlp , do you mind reviewing this change? Thanks

**Comment by jiqing-feng** [comment]
Hi @Titus-von-Koeller . This change is needed for bitsandbytes cpu path, can you help to review it?

Besides, it's also needed for AWQ cpu path which is already enabled here: #33460

**Comment by jiqing-feng** [comment]
The following script can reproduce the AWQ error:

```python
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, AwqConfig

model_id = "PrunaAI/JackFram-llama-68m-AWQ-4bit-smashed"

text = ["I am happy because", "This is"]
tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token
input_ids = tokenizer(text, return_tensors="pt", padding=True)

# quantization_config = BitsAndBytesConfig(load_in_8bit=True)
quantization_config = AwqConfig(version="ipex")

model = AutoModelForCausalLM.from_pretrained(model_id, device_map="auto", quantization_config=quantization_config)
model.generation_config.cache_implementation = "static"
model.generate(**input_ids)

```

**Comment by jiqing-feng** [comment]
Hi @aymeric-roucher  @LysandreJik , do you have time to review this bug fix? Thanks!

**Comment by zucchini-nlp** [comment]
@jiqing-feng sorry for late reply, the `transformers` team was off last week and someone will review the PR soon. Seems like you tagged all relevant people already

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1635,7 +1635,8 @@ def get_layer_device_map(execution_device_map: Optional[dict] = None):
             # This is needed here if we don't want to make changes in accelerate in order to save execution_device
             # For offloaded case, we need to get the execution device, not just the device where it is offloaded
             if hasattr(self, "hf_device_map"):
-                main_device = [d for d in self.hf_device_map.values() if d not in ["cpu", "disk"]][0]
+                main_device = [d for d in self.hf_device_map.values() if d not in ["cpu", "disk"]]
+                main_device = main_device[0] if main_device else "cpu"
```
Comment (SunMarc) [inline src/transformers/generation/utils.py:1639]: What about this ? This is taken from the `dispatch_model` function. 
```suggestion
                if set(self.hf_device_map.values()) == {"cpu"} or set(self.hf_device_map.values()) == {"cpu", "disk"}:
                    main_device = "cpu"
                else:
                    main_device = [d for d in self.hf_device_map.values() if d not in ["cpu", "disk"]][0]
```

**Review by SunMarc — APPROVED** [review]
Thanks for the fix ! I've suggested a similar fix that follows more closely what we have in accelerate. LMK if this works for you !

**Comment by jiqing-feng** [comment]
Hi @SunMarc , I have applied your changes, thanks!

**Review by SunMarc — APPROVED** [review]
Nice, thanks !

**Review by ArthurZucker — APPROVED** [review]
LGTM 🤗

**Pull request merged**